### PR TITLE
feat: SPL compression, decompression, refactors

### PR DESCRIPTION
**Pull request by SwenSchaeferjohann**
* refactored: JS c-token transfer and ix packing
* added: compress,decompress SPL ixs, actions, cli commands, test coverage (happy paths)
* adjusted programs to support compression, token, `mint` field into `create_transfer_instruction`() to support zero-token-inputs (e.g. for spl-compression)

note: locally we had to run with `cargo test-sbf -- --test-threads=1` otherwise we'd get gnark-server contention.

### CODE REVIEW

File: programs/compressed-token/src/process_transfer.rs [hunk 1]
```diff
@@ -33,16 +34,21 @@ pub fn process_transfer<'a, 'b, 'c, 'info: 'b + 'c>(
             &ctx.accounts.authority.key(),
         )?;
 
+    // TODO: check if this is already implemented
```
Comment (SwenSchaeferjohann) [inline programs/compressed-token/src/process_transfer.rs:37]: this was a remnant from rebasing spl-compression onto the latest main: the delegation check is already implemented via `get_input_compressed_accounts_with_merkle_context_and_check_signer` correct? @ananas-block

File: programs/compressed-token/src/process_transfer.rs [hunk 2]
```diff
@@ -456,9 +479,13 @@ pub mod transfer_sdk {
         leaf_indices: &[u32],
         proof: &CompressedProof,
         input_token_data: &[crate::TokenData],
+        mint: Pubkey,
```
Comment (SwenSchaeferjohann) [inline programs/compressed-token/src/process_transfer.rs:482]: FYI: we have to pass mint separately because in case of spl-compress, input_token_data is empty

File: programs/compressed-token/src/process_transfer.rs (hunk 1 above)
Comment (ananas-block) [reply programs/compressed-token/src/process_transfer.rs:37]: Yes delegate checks are implemented but untested.
Also we don't have an instruction to create a compressed token account with a delegate yet.

File: programs/compressed-token/tests/test.rs [hunk 3]
```diff
@@ -1,5 +1,7 @@
 #![cfg(feature = "test-sbf")]
 
+use std::{print, println};
```
Comment (ananas-block) [inline programs/compressed-token/tests/test.rs:3]: why is this necessary?

File: programs/compressed-token/src/process_mint.rs [hunk 4]
```diff
@@ -265,14 +268,15 @@ pub mod mint_sdk {
...
-    use crate::{get_token_authority_pda, get_token_pool_pda};
+    use crate::{get_cpi_authority_pda, get_token_authority_pda, get_token_pool_pda};
 
     pub fn create_initialize_mint_instruction(
         fee_payer: &Pubkey,
         authority: &Pubkey,
         mint: &Pubkey,
     ) -> Instruction {
         let token_pool_pda = get_token_pool_pda(mint);
+        println!("token_pool_pda: {:?}", token_pool_pda);
```
Comment (ananas-block) [inline programs/compressed-token/src/process_mint.rs:279]: leftover

File: programs/compressed-token/src/process_transfer.rs [hunk 5]
```diff
@@ -557,6 +584,8 @@ pub mod transfer_sdk {
             pubkey_array,
```
Comment (ananas-block) [inline programs/compressed-token/src/process_transfer.rs:584]: pubkey array has been deleted from this struct onchain but still exists here.
Happy to add the accounts to remaining accounts shouldn't be a big change.

File: js/compressed-token/src/actions/decompress.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        outputStateTree: merkleTree,
+        recentInputStateRootIndices: proof.rootIndices,
+        recentValidityProof: proof.compressedProof,
+    });
+
+    const { blockhash } = await rpc.getLatestBlockhash();
+    const additionalSigners = dedupeSigner(payer, signers);
+    const signedTx = buildAndSignTx(ixs, payer, blockhash, additionalSigners);
+    const txId = await sendAndConfirmTx(rpc, signedTx, confirmOptions);
+    return txId;
```
Comment (ananas-block) [inline js/compressed-token/src/actions/decompress.ts:87]: on a general note, we could think about returning more information here such as which compressed accounts have been spent or created etc.

File: js/compressed-token/src/actions/transfer.ts [hunk 7]
```diff
@@ -101,56 +58,70 @@ export async function transfer(
...
-        ),
-        [merkleTree, merkleTree], // out state trees
-        inputAccounts.map(account => account.compressedAccount), // input compressed accounts
-        [recipientCompressedAccount, changeCompressedAccount], // output compressed accounts
-        // TODO: replace with actual recent state root index!
-        // This will only work with sequential state updates and no cranking!
-        proof.rootIndices, // input state root indices
-        proof.compressedProof,
-    );
+    const ixs = await CompressedTokenProgram.transfer({
```
Comment (ananas-block) [inline js/compressed-token/src/actions/transfer.ts:70]: Naming is confusing to me ``createTransferInstruction`` is more intuitive in my opinion.
The name transfer leads me to expect for a transaction being sent not instructions being returned.

File: js/compressed-token/src/program.ts [hunk 8]
```diff
@@ -4,27 +4,137 @@ import {
...
+    /**
+     * The payer of the transaction.
+     */
+    payer: PublicKey;
+    /**
+     * owner of the *uncompressed* token account.
+     */
+    owner: PublicKey;
+    /**
+     * source (associated) token account address.
```
Comment (ananas-block) [inline js/compressed-token/src/program.ts:49]: actual address of the ata or the public key the ata is derived from?

File: js/compressed-token/src/program.ts [hunk 9]
```diff
@@ -4,27 +4,137 @@ import {
...
+     * The recent state root indices of the input state. The expiry is tied to
+     * the proof.
+
+     */
+    recentInputStateRootIndices: number[];
+    /**
+     * The recent validity proof for state inclusion of the input state. It
+     * expires after n slots.
+     */
+    recentValidityProof: CompressedProof;
```
Comment (ananas-block) [inline js/compressed-token/src/program.ts:130]: nice the recent makes a lot of sense, we should add an error for validity proof expired!

Comment (ananas-block) [reply js/compressed-token/src/program.ts:130]: instead of the root not found error.

File: js/compressed-token/src/actions/decompress.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+    const proof = await rpc.getValidityProof(
+        inputAccounts.map(account => bn(account.compressedAccount.hash)),
+    );
+
+    const ixs = await CompressedTokenProgram.decompress({
+        payer: payer.publicKey,
+        inputCompressedTokenAccounts: inputAccounts,
+        toAddress: toAddress, // TODO: add check that it is a token account
+        amount: amount,
```
Comment (ananas-block) [inline js/compressed-token/src/actions/decompress.ts:77]:
```suggestion
        toAddress, // TODO: add check that it is a token account
        amount,
```

File: js/compressed-token/src/actions/compress.ts [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+    amount = bn(amount);
+
+    const ixs = await CompressedTokenProgram.compress({
+        payer: payer.publicKey,
+        owner: currentOwnerPublicKey,
+        source: sourceTokenAccount,
+        toAddress: toAddress,
+        amount: amount,
+        mint: mint,
```
Comment (ananas-block) [inline js/compressed-token/src/actions/compress.ts:62]:
```suggestion
        toAddress,
        amount,
        mint,
```

File: js/compressed-token/src/actions/compress.ts [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    mint: PublicKey,
+    amount: number | BN,
+    owner: Signer | PublicKey,
+    sourceTokenAccount: PublicKey,
+    toAddress: PublicKey,
+    merkleTree: PublicKey = defaultTestStateTreeAccounts().merkleTree,
+    multiSigners: Signer[] = [],
+    confirmOptions?: ConfirmOptions,
+): Promise<TransactionSignature> {
+    const [currentOwnerPublicKey, signers] = getSigners(owner, multiSigners);
```
Comment (ananas-block) [inline js/compressed-token/src/actions/compress.ts:52]: I am a bit confused by multiple signers and separate payer and owner, but need to think about how I would do it differently.

File: js/compressed-token/src/program.ts [hunk 13]
```diff
@@ -130,7 +344,10 @@ export class CompressedTokenProgram {
     /** @internal */
     static get deriveCpiAuthorityPda(): PublicKey {
         const [address, _] = PublicKey.findProgramAddressSync(
-            [Buffer.from('cpi_authority')],
+            [
+                Buffer.from('cpi_authority'),
```
Comment (ananas-block) [inline js/compressed-token/src/program.ts:348]: what about creating a constant for this or even taking the constant from the idl?

**Review by ananas-block — COMMENTED** [review]
Awesome, looks good in general.
Couple small comments.

File: programs/compressed-token/tests/test.rs (hunk 3 above)
Comment (SwenSchaeferjohann) [reply programs/compressed-token/tests/test.rs:3]: removed unnecessary prints

File: js/compressed-token/src/program.ts (hunk 9 above)
Comment (SwenSchaeferjohann) [reply js/compressed-token/src/program.ts:130]: yea we can add a method that explicitly validates the recency, but i think `root not found` is pretty decent for a start. (equivalent to `blockhash not found`); client can reason that it's most likely bc expired.

File: js/compressed-token/src/program.ts (hunk 13 above)
Comment (SwenSchaeferjohann) [reply js/compressed-token/src/program.ts:348]: made constant

File: js/compressed-token/src/actions/compress.ts (hunk 12 above)
Comment (SwenSchaeferjohann) [reply js/compressed-token/src/actions/compress.ts]: yea multiSigners was a leftover from dealing with multisigs, (i.e. `mintAuthority` and `owner` being a multisig)
we'll have to deal with that in a future pr - straightforward support for multisigs.

File: js/compressed-token/src/actions/transfer.ts (hunk 7 above)
Comment (SwenSchaeferjohann) [reply js/compressed-token/src/actions/transfer.ts:66]: yea agree it's more descriptive however .transfer() matches the solana systemprogram implementation, so there's familiarity for folks who work with it e.g. in browser environments...

File: js/compressed-token/src/actions/decompress.ts (hunk 6 above)
Comment (SwenSchaeferjohann) [reply js/compressed-token/src/actions/decompress.ts:83]: yeah good idea, noted for a followup pr

**Review by vadorovsky — APPROVED** [review]
